pkg/function/service.go

```diff
@@ -81,26 +117,47 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.
 // ReduceFn applies a reduce function to a datum stream.
 func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer) error {
 	var (
-		ctx      = stream.Context()
-		key      string
-		reduceCh = make(chan Datum)
-		md       Metadata
+		key       string
+		md        Metadata
+		err       error
+		startTime int
+		endTime   int
+		reduceCh  = make(chan Datum)
+		ctx       = stream.Context()
 	)
 
-	// get key and metadata from gPRC metadata
-	if grpcMD, ok := metadata.FromIncomingContext(ctx); ok {
-		// get Key
-		keyValue := grpcMD.Get(DatumKey)
-		if len(keyValue) > 1 {
-			return fmt.Errorf("expect extact one key but got %d keys", len(keyValue))
-		} else if len(keyValue) == 1 {
-			key = keyValue[0]
-		} else {
-			// do nothing: the length equals zero is valid, meaning the key is an empty string ""
-		}
-		// TODO: get metadata
+	grpcMD, ok := grpcmd.FromIncomingContext(ctx)
+	if !ok {
+		return fmt.Errorf("key and window information are not passed in grpc metadata")
+	}
+
+	// get key from gPRC metadata
+	key, err = getValueForKey(grpcMD, DatumKey)
+	if err != nil {
+		return err
+	}
+
+	// get window start and end time from grpc metadata
+	var st, et string
+	st, err = getValueForKey(grpcMD, WinStartTime)
+	if err != nil {
+		return err
+	}
+
+	et, err = getValueForKey(grpcMD, WinEndTime)
+	if err != nil {
+		return err
 	}
 
+	startTime, _ = strconv.Atoi(st)
+	endTime, _ = strconv.Atoi(et)
+
+	// create interval window interface using the start and end time
+	iw := NewIntervalWindow(time.UnixMilli(int64(startTime)), time.UnixMilli(int64(endTime)))
+
+	// create metadata using interval window interface
+	md = NewMetadata(iw)
```

Comment:
do we need Metadata? why dont we make it explicit and pass in IntervalWindow as an argument?
we dont expect any additional metadata here.

Reply:
I prefer Metadata. If we are to have more metadata fields in the future, we can directly add in the `md` and the user doesn't have to update their reduce handler signature. e.g. https://github.com/numaproj/numaflow-go/blob/main/pkg/function/examples/sum/main.go#L12